cityscapesscripts/evaluation/evalObjectDetection3D.py

```diff
@@ -0,0 +1,1186 @@
+
+        logger.info("Loading predictions...")
+        predictions = getFiles(pred_folder)
+
+        predictions.sort()
+        logger.info("Found %d prediction files." % len(predictions))
+
+        for p in predictions:
+            preds_for_image = []
+
+            # extract CITY_RECORDID_IMAGE from filepath
+            base = os.path.basename(p)
+
+            base = base[:base.rfind("_")]
+
+            with open(p) as f:
+                data = json.load(f)
+
+            for d in data["annotation"]:
+                if (
+                    "class_name" in d.keys() and
+                    d["class_name"] in self.eval_params.labels_to_evaluate
+                ):
+                    try:
+                        box_data = Box3DObject(d)
+                    except:
+                        logger.critical("Found incorrect annotation in %s." % p)
+                        continue
+
+                    preds_for_image.append(box_data)
+
+            self.preds[base] = {
+                "objects": preds_for_image
+            }
+
+    def evaluate(self) -> None:
+        """Main evaluation routine."""
+
+        # fill up predictions dict with empty detections if prediction file not found
+        for base in self.gts.keys():
+            if base not in self.preds.keys():
+                logger.critical(
+                    "Could not find any prediction for image %s." % base)
+                self.preds[base] = {"objects": []}
+
+        # initialize empty data
+        for s in self._conf_thresholds:
+            self._stats[s] = {
+                "data": {}
+            }
+
+        logger.info("Evaluating images...")
+        # calculate stats for each image
+        self._calcImageStats()
+
+        logger.info("Calculate AP...")
+        # calculate 2D ap
+        self._calculateAp()
+
+        logger.info("Calculate TP stats...")
+        # calculate FP stats (center dist, size similarity, orientation score)
+        self._calcTpStats()
+
+    def saveResults(self, result_folder: str) -> str:
+        """Saves the evaluation results to ``"results.json"``
+
+        Args:
+            result_folder (str): directory in which the result files are saved
+
+        Returns:
+            str: filepath of ``"results.json"``
+        """
+
+        result_file = os.path.join(result_folder, "results.json")
+        with open(result_file, 'w') as f:
+            json.dump(self.results, f, indent=4)
+
+        # dump internal stats for debugging
+        # stats_file = os.path.join(result_folder, "stats.json")
+        # with open(stats_file, 'w') as f:
+        #    json.dump(self._stats, f, indent=4)
+
+        return result_file
+
+    def _calcImageStats(self) -> None:
+        """Internal method that calculates Precision and Recall values for whole dataset."""
+
+        # single threaded
+        results = []
+        for x in tqdm(self.gts.keys()):
+            results.append(self._worker(x))
+
+        # multi threaded
```

Comment:
So multi thread does not work? Maybe we should delete the code then?

Reply:
Removed multithreading.